### PR TITLE
Compatible navigation manager + more consistent popstate

### PR DESCRIPTION
**Pull request by daniel-sc**
The added `UriFragmentNavManager` creates URLs that are compatible with the default Vaadin `UriFragmentManager`.

As browser behaviour for `popstate` is inconsistent, I fired events manually for `pushState`/`replaceState`. So  no URL change will be missed. Duplicate events are recognised by equal URL and are dismissed.

### CODE REVIEW

**Comment by daniel-sc** [comment]
Hi @wolfie,
it would be nice if you could merge this pull request and cut a new release.
Best, Daniel

**Comment by wolfie** [comment]
Thanks for your contribution. Seriously, I mean it.

I feel bad, since I'm quite lazy at maintaining my add-ons. Actually, I've
been a Scrum Master for the past 1.5 years, so I haven't touched any code
for ages and my dev env is long gone.

I have a slew other excuses that would make me sound busy and stuff, but
honestly, I don't know what to say. I'll try to gather the focus and
review _your
totally awesome contribution_ this weekend, and if (probably when) I don't
see anything off kilter, I'll just merge and re-publish.

Thanks, again.

File: src/main/java/com/github/wolfie/history/HistoryExtension.java [hunk 1]
```diff
@@ -695,7 +634,7 @@ protected void callFunction(final String name, final Object... arguments) {
 
     public NavigationStateManager createNavigationStateManager(
             final String urlRoot) {
-        return new NavManager(urlRoot);
+        return new NavManager(this, urlRoot);
```
Comment (wolfie) [inline src/main/java/com/github/wolfie/history/HistoryExtension.java:637]: What was your reason to move the private inner class to a package-private external class?

**Review by wolfie — COMMENTED** [review]
Ok, this pull request is confusing me a bit. One part is simply me being out of touch of what the existing code does. But another problem is that there are many things being changed in a single pull request. Also, the use case for this solution is unclear to me. Would you mind telling me the problem that you were having, and why this solution worked for you?

For example, `com.vaadin.navigator.Navigator.UriFragmentManager` seems to be a Vaadin-internal class, that, according to its JavaDoc, is only exposed due to testing limitations. Your implementation of `UriFragmentNavManager` extends that, which makes me wonder if this is something that should be done in the first place. Is this something similar that `HistoryExtension.configurePushStateEnabledNavigator` tries to solve for the user?

**Comment by daniel-sc** [comment]
Hi @wolfie,

thanks very much for the specific feedback! I'll try to explain, but if that (in some parts) does not convince you, I'll be happy to update according your guidline.

**Use Case**
Our application currently uses standard `com.vaadin.navigator.Navigator` (which entails `com.vaadin.navigator.Navigator.UriFragmentManager`). We want to use history api instead (to be able to manage URL query additionally to URL fragment). Furthermore we'd like to keep existing URLs - i.e. the (new) `NavigationStateManager` should use the same state URL-encoding as existing `UriFragmentManager` (e.g. `http://localhost:8080/#!my_view_name`).

**Extending `UriFragmentManager`**
Good point about this class beeing ("mostly" ;) ) interal for Vaadin.
On the other hand: `UriFragmentNavManager` shares behavior (same encoding of state), but only uses different implementation ("persistence" via pushState). I think it is fine that `NavManager` does _not_ extend `UriFragmentManager`, as it encodes state differently.

**Refactoring `NavManager` to separate file**
This was just to unclutter `HistroyExtension.java`. You are right, that this makes the class (slightly) more visible, but I'd even go for the `public` qualifier to allow users to extend/change `NavManager`. As the HistoryExtension can/will be used with different `NavigationStateManager`s, I didn't include further factory methods (`HistoryExtension.configurePushStateEnabledNavigator`), but instead let the user create the `UriFragmentNavManager`.

**Updated `popstate` listener**
When testing with IE11, I found that it did not send `popstate` events when changing URL via `pushState` but Chrome/Firefox did (opposed to what is documented e.g. here: https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onpopstate ). To be browser agnositic, the event is immediately fired (server-side) when `pushState`/`replaceState` is invoked. To prevent duplicate events, any event is checked against the URL from the previous event and duplicates are discarded.

Does that clarify things?

**Comment by daniel-sc** [comment]
Hi @wolfie,

I suppose you have been to busy to look into this yet - or is there something elso blocking you?
As I said, if you like me to change things, please come forward - I'll be cooperative ;-)

Best, Daniel

**Comment by wolfie** [comment]
Thank you @daniel-sc!

Yes, you are right in assuming that I've been preoccupied. I recently changed jobs, I'm basically internetless and I'm using a fresh computer without anything installed (I wish I was making this up) for the next couple of weeks. Nevertheless, it's not your problem, and I owe you a reply. Doubly so, with your awesome reply. Thank you for taking the time.

I'm not doubting your use case, but I always get weirded out if someone ends up using API that is intended as internal. It sounds like Vaadin isn't supporting your use case properly. @legioth, would you have the time to look into this and share some thoughts?

**Comment by daniel-sc** [comment]
@wolfie if it would make you more comfortable, I can as well update `UriFragmentNavManager` to _not_ extend Vaadin `UriFragmentManager`. Should I?
I'm more into getting this merge request trough, than making a point ;-)

**Comment by wolfie** [comment]
I'm more worried that the framework has something that should be fixed, and thought that Leif would have some input on this, rather than worked around in addons. 

In any case, my hesitation is mostly coming from the fact that I'm nowadays so alienated of Vaadin code. I made you a collaborator for this repository, so feel free to do what you feel is best :). You obviously know more than I do these days.

If you want, I can either publish the add-on in the Directory, or I can take my copy down, and you can continue publishing it...

**Comment by daniel-sc** [comment]
@wolfie thanks for merging! 
It would be great if you publish the new version - this is probably easier than creating a 'new' add-on.. But if you can't (because of missing setup or something else), I can of course as well publish it

**Comment by wolfie** [comment]
No problem, thanks for the code @daniel-sc :)

I'll see if I can get the setup up quickly and try to update it. I'll let you know how it goes

**Comment by wolfie** [comment]
it is done https://vaadin.com/directory#!addon/history

**Pull request merged**